Remove old File Upload article + improve the new one

wouterj opened the pull request:
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | yes |
| Applies to | 2.3+ |
| Fixed tickets | #5375 |

The old file upload article wasn't good, a new one was written by @javiereguiluz but the old one remained online. This PR removes the old one and documents the missing bits in the new article.

wouterj committed:
Include PHP templating code examples

wouterj committed:
Include better example using a parameter

wouterj committed:
Some minor fixes

wouterj committed:
Document extracting upload to a service

wouterj committed:
Document using a Doctrine listener

wouterj committed:
Delete old file upload article

WouterJ commented:
Btw, why do we use an Entity property to store both the uploaded file and file path? Wouldn't it be better to use an unmapped field?

WouterJ commented on cookbook/controller/upload_file.rst:
```diff
@@ -164,15 +187,203 @@ There are some important things to consider in the code of the above controller:
+
+If you are using Doctrine to store the Product entity, you can create a
+:doc:`Doctrine listener </cookbook/doctrine/event_listeners_subscribers>` to
+automatically upload the file when persisting the entity::
```
I wondering if we should show code + config, or just link to the other article and just mentioning this as a little tip.

weaverryan replied:
I like how you've handled it here: with code+config (it's really not too long).

WouterJ commented on cookbook/controller/upload_file.rst:
```diff
@@ -164,15 +187,203 @@ There are some important things to consider in the code of the above controller:
+Creating an Uploader Service
+----------------------------
+
+To avoid logic in controllers, making them big, you can extract the upload
```
Not sure if I really like this service, it's small doesn't do much and doesn't seem to make things simpler in any way. However, it felt like a nice step between a big controller and using a Doctrine listener.

Raistlfiren commented:
@WouterJ, if I might make an addition as well.... It might be good to include how to handle files when editing them. For example, if I have a simple form like you included in the example, then I would need to either set data_class to null on the brochure field or convert the string to a File object. Otherwise, the form will error out and say - "The form's view data is expected to be an instance of class Symfony\Component\HttpFoundation\File\File, but is a(n) string". Maybe creating the file object is better in a data transformer, not sure. Thoughts?

xabbuh commented:
> Btw, why do we use an Entity property to store both the uploaded file and file path? Wouldn't it be better to use an unmapped field?

I think this was done to always trigger the lifecycle callback methods. If only the file path were mapped to the database, lifecycle callbacks would never be triggered if only the uploaded file was changed (Doctrine would not treat the entity as changed at all).

However, I tend to say that we should rather move the whole file uploding logic into a DTO and not pollute the entity with logic that doesn't belong there.

weaverryan commented on cookbook/controller/upload_file.rst:
```diff
@@ -146,16 +159,26 @@ Finally, you need to update the code of the controller that handles the form::
+    # app/config/parameters.yml
+    parameters:
+        # ...
+        brochures_directory: '%kernel.root_dir%/../web/uploads/brochures'
```
This makes more sense in `config.yml` - it's not a value that would normally need to be changed on a machine-by-machine basis, and devs wouldn't/shouldn't need to worry about configuring it.

weaverryan replied:
And actually, we're dependent on it later when we make the paths to it in the Twig template.

weaverryan commented on cookbook/controller/upload_file.rst:
```diff
@@ -164,15 +187,203 @@ There are some important things to consider in the code of the above controller:
+            $fileName = $this->uploader->upload($file);
+
+            $entity->setBrochure($fileName);
+        }
```
Can we add the preUpdate too? That would be more useable

weaverryan commented:
@WouterJ I added 2 small comments, but :+1: from me after that. Sorry for the long delay - I'm also excited to remove the old, terrible article.

Koc commented:
@WouterJ please add info about uploading and validating multiple files, see #5603 request

Koc commented:
@weaverryan can we create redirect from old url to the new article?

xabbuh commented:
@Koc That's done with the entry in the `redirection_map` file.

wouterj committed:
Updated after Ryan's review

WouterJ commented:
I've updated the article. Should be ready to merge now

javiereguiluz commented on cookbook/controller/upload_file.rst:
```diff
@@ -164,15 +188,220 @@ There are some important things to consider in the code of the above controller:
+        }
+    }
+
+This class can be registered as a service in the service container:
```
`This class can be registered as a service in the service container:`

->

`Then, define a service for this class:` ?

javiereguiluz commented on cookbook/controller/upload_file.rst:
```diff
@@ -164,15 +188,220 @@ There are some important things to consider in the code of the above controller:
+        }
+    }
+
+Now, register this class as a Doctrine listener:
```
I have a question about this. In the popular VichUploaderBundle, they warn you about uploading files in entities which haven't modified any other property. In that case, it won't work and the file will be lost. That's why you need to fake some changes (e.g. updating the `updatedtAt` property). Would this be a problem here too?

WouterJ replied:
Hmm, will have to look into this. I think it might is (as doctrine doesn't discover the entity changed)

xabbuh replied:
If I understand the code correctly, the file name will always change when uploaded thus leading to changes in the entity which means that Doctrine is able to detect the change.

WouterJ replied:
Created a test project and this is not a problem with this code.

wouterj committed:
Few fixes after testing the code

wouterj committed:
Add two tips about converting to File instance after fetching

wouterj committed:
Less verbose sentence, thanks Javier!

WouterJ commented:
While testing this code in an application I found a few bugs which I fixed in f6068c9. Also, I strumbled onto issues creating an edit form as it needs a `File` instance. To avoid problems for readers, I've added two little tip boxes in 0c54b8a.

It's now even more ready to merge than it was this afternoon :)

javiereguiluz commented:
@WouterJ amazing! Thanks for testing everything thoroughly in a real project 🙇

xabbuh commented on cookbook/controller/upload_file.rst:
```diff
@@ -164,15 +188,268 @@ There are some important things to consider in the code of the above controller:
-   simplifies the code: ``$this->container->getParameter('brochures_dir')``.
 
-You can now use the following code to link to the PDF brochure of an product:
+You can use the following code to link to the PDF brochure of an product:
```
a product

xabbuh commented:
👍 I like the changes you made here.

However, for the future I think we can improve things even more by introducing a DTO that would be used to build the form and doing the file upload. That way we would not need to have the hacks being present currently.

weaverryan commented:
I left a minor comment, but let's merge this :) 👍

WouterJ commented:
Updated again. @xabbuh using a DTO is exactly what I thought when improving this article.

@weaverryan I can't find your minor comment.

javiereguiluz commented:
@WouterJ you say "updated again" but I can't see any new commits 😕

WouterJ commented:
@javiereguiluz as it was a very small typo fix, I merged it into the previous commit: https://github.com/symfony/symfony-docs/pull/6040/commits/94e21ff80bff77063efc0924ff84331c2ad7250c#diff-bcae3aaf74848d919ee531cad0d95547R192

xabbuh commented:
Thank you Wouter.
